test/topology_custom/test_truncate_with_tablets.py

```diff
@@ -0,0 +1,55 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+from cassandra.query import SimpleStatement, ConsistencyLevel
+from test.pylib.manager_client import ManagerClient
+from test.topology.conftest import skip_mode
+import pytest
+import logging
+import asyncio
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_truncate_while_migration(manager: ManagerClient):
+
+    logger.info('Bootstrapping cluster')
+    cfg = { 'enable_tablets': True,
+            'error_injections_at_startup': ['migration_streaming_wait']
+            }
+
+    servers = []
+    servers.append(await manager.server_add(config=cfg))
+
+    cql = manager.get_cql()
+
+    # Create a keyspace with tablets and initial_tablets == 2, then insert data
+    await cql.run_async("CREATE KEYSPACE test WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} AND tablets = {'initial': 2}")
+    await cql.run_async('CREATE TABLE test.test (pk int PRIMARY KEY, c int);')
```

Comment:
Please add injection-based coverage for a concurrent DROP TABLE of a table that's being truncated. 
A concurrent repair has been mentioned elsewhere, will it be addressed in a separate issue?

Reply:
Another part of the test plan is executing TRUNCATE with a down node. 
1) Start TRUNCATE
2) Block it on an injection against one of the nodes.
3) Down that node.
4) Bring the node back up
5) (Optionally) drop the table somewhere between 3) and 6)
6) Follow up on truncate completion - it should succeed.

Reply:
> Please add injection-based coverage for a concurrent DROP TABLE of a table that's being truncated.

I will add that in the next version.

> A concurrent repair has been mentioned elsewhere, will it be addressed in a separate issue?

I think serializing repair with migrations and other topology operations has already been addressed: #18641 
Because repair works by holding the ERM pointer and therefor blocking topology changes, it will also avoid concurrent truncate, because the topology coordinator can not advance until ERM is released.

Reply:
> Another part of the test plan is executing TRUNCATE with a down node.
> 
> 1. Start TRUNCATE
> 2. Block it on an injection against one of the nodes.
> 3. Down that node.
> 4. Bring the node back up
> 5. (Optionally) drop the table somewhere between 3) and 6)
> 6. Follow up on truncate completion - it should succeed.

I agree. Will add this in the next version.

Reply:
I think the ERM magic raises concerns of TRUNCATE starvation, because repairs don't wait for each other, and TRUNCATE has to wait for all repairs to finish. @tgrabiec what am I missing?

Reply:
@nodep Holding erm will block truncate only if repair started earlier (because truncate executes a barrier). It won't prevent later repair from running because repair checks for tablet transitions, not topology ops, and truncate doesn't issue tablet transitions.

```
        while (true) {
            _repair_module->check_in_shutdown();
            erm = t->get_effective_replication_map();
            const locator::tablet_map& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
            if (!tmap.has_transitions()) {
                break;
            }
            rlogger.info("repair[{}] Table {}.{} has tablet transitions, waiting for topology to quiesce", rid.uuid(), keyspace_name, table_name);
            erm = nullptr;
            co_await container().invoke_on(0, [] (repair_service& rs) {
                return rs._tsm.local().await_not_busy();
            });
            rlogger.info("repair[{}] Topology quiesced", rid.uuid());
        }
```

Yes, repair currently can block truncate for a long time due to high granularity. 

It's less of a problem with https://github.com/scylladb/scylladb/pull/21088 which moves repair to tablet scheduler. With that method, truncate request will only wait for currently repaired tablets to finish, it will prevent new tablet repairs from being started (pending request = preempt scheduling). 

Those erm hacks were supposed to be temporary. We should move nodetool repair to tablet scheduler. @asias says there are no such plans. I think it's a mistake. @mykaul @bhalevy @avikivity

Reply:
Let's proceed with the understanding that #21088 is going to be merged.

I agree nodetool repair should use the scheduler, at least for the simple cases.

Reply:
> Let's proceed with the understanding that #21088 is going to be merged.
> 
> I agree nodetool repair should use the scheduler, at least for the simple cases.

I concur.
Even if providing a token to repair, we can schedule repair only for the tablet containing the requested token.

Reply:
> > Let's proceed with the understanding that #21088 is going to be merged.
> > I agree nodetool repair should use the scheduler, at least for the simple cases.
> 
> I concur. Even if providing a token to repair, we can schedule repair only for the tablet containing the requested token.

As for the other repair options, like limiting the dc or the participating nodes, we may want to support only some of them for tablets.  limiting to a source dc does make sense also in the automatic repair framework (we can even schedule tablet repair per DCs independently), but I'm not sure about limiting to particular hosts.

Reply:
Test cases for concurrent DROP TABLE and node restart have been added in v7.